### ISSUE TITLE
Helm3-beta2: Upgrade does not change image version in deployment

### ISSUE THREAD

**Issue by jkrayl** [helm/helm#6358]
The image version of a deployment doesn't change after I up/downgrade to a newer/older chart version with a different appVersion.

e.g.:
Output of `helm ls`:
```
NAME    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART
mongo   mongo           1               2019-09-05 15:52:16.95821 +0200 CEST    deployed        mongodb-7.2.7
```
```
helm upgrade  mongo stable/mongodb --version 6.2.0
```
`helm ls`
```
NAME    NAMESPACE       REVISION        UPDATED                                 STATUS          CHART        
mongo   mongo           2               2019-09-05 15:54:14.122248 +0200 CEST   deployed        mongodb-6.2.0
```
After I downgraded mongodb from chart version 7.2.7 to 6.2.0 the chart version changed but the deployment still used the old image version. (4.0.12 instead of 4.0.10)

Output of `kubectl get deployments mongo-mongodb -o yaml | grep image`: 
```
        image: docker.io/bitnami/mongodb:4.0.12-debian-9-r22
```
Output of `kubectl get deployments mongo-mongodb -o yaml | grep chart` on same deployment:
```
    chart: mongodb-6.2.0
        chart: mongodb-6.2.0
```
Same issue appeared with different charts like grafana.

Output of `helm version`:
```
version.BuildInfo{Version:"v3.0.0-beta.2", GitCommit:"26c7338408f8db593f93cd7c963ad56f67f662d4", GitTreeState:"clean", GoVersion:"go1.12.9"}
```
Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T12:36:28Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```
Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
`AKS`

**Comment by Tarkleigh** [comment]
I also have this problem when using Helm3 beta3. While the chart looks upgrade, the image used inside the deployment does not change and always uses the version that was originally deployed. It seems like Helm3 is not processing updates correctly and does not even notice anything went wrong.

Output of helm version:

> version.BuildInfo{Version:"v3.0.0-beta.3", GitCommit:"5cb923eecbe80d1ad76399aee234717c11931d9a", GitTreeState:"clean", GoVersion:"go1.12.9"}


Output of kubectl version

> Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T12:36:28Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
> Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider

> AKS

**Comment by bacongobbler** [comment]
Marking as a bug to try and recreate.

**Comment by karuppiah7890** [comment]
I'm able to reproduce the issue with these two charts

[issue-6358-chart.zip](https://github.com/helm/helm/files/3587837/issue-6358-chart.zip)

[issue-6358-chart-new.zip](https://github.com/helm/helm/files/3587836/issue-6358-chart-new.zip)

The diff between the two is, the first version has image tag as `1.17` and the second one has image tag as `stable`

```
$ helm upgrade --install dummy issue-6358-chart

$ kubectl get deploy dummy-dummy-deployment-chart -o yaml | rg image
      - image: nginx:1.17
        imagePullPolicy: IfNotPresent

$ helm upgrade --install dummy issue-6358-chart-new/

$ kubectl get deploy dummy-dummy-deployment-chart -o yaml | rg image
      - image: nginx:1.17
        imagePullPolicy: IfNotPresent
```

Weirdly when I check the values I see that it hasn't changed. Also, it looks different from the values that I have seen in helm v2 (v2 only shows user supplied values, with no default values yaml I guess? and then shows "computed values" on debug) . Will have to check what's going wrong

```
$ helm history dummy
REVISION        UPDATED                                 STATUS          CHART                           APP VERSIONDESCRIPTION
1               2019-09-08 23:07:30.638268 +0530 IST    superseded      dummy-deployment-chart-0.1.0    1.0        Install complete
2               2019-09-08 23:07:36.259785 +0530 IST    superseded      dummy-deployment-chart-0.2.0    1.0        Upgrade complete

$ helm get values dummy --revision 1
affinity: {}
fullnameOverride: ""
image:
  pullPolicy: IfNotPresent
  repository: nginx
  tag: 1.17
imagePullSecrets: []
ingress:
  annotations: {}
  enabled: false
  hosts:
  - host: chart-example.local
    paths: []
  tls: []
nameOverride: ""
nodeSelector: {}
replicaCount: 1
resources: {}
service:
  port: 80
  type: ClusterIP
tolerations: []

$ helm get values dummy --revision 2
affinity: {}
fullnameOverride: ""
image:
  pullPolicy: IfNotPresent
  repository: nginx
  tag: 1.17
imagePullSecrets: []
ingress:
  annotations: {}
  enabled: false
  hosts:
  - host: chart-example.local
    paths: []
  tls: []
nameOverride: ""
nodeSelector: {}
replicaCount: 1
resources: {}
service:
  port: 80
  type: ClusterIP
tolerations: []
```

**Comment by karuppiah7890** [comment]
Feels like helm3 shows release values similar to helm2, as I see a `--all` flag in `helm get values` command to see the computed values. So, normally just user supplied values should be shown.

I guess the issue is in the part where I see the values even for revision 1 when I never passed any values. And that caused the problem during upgrade where the existing values got copied as new release has no value and existing release had values due to the install issue. This copy happens due to this code

https://github.com/helm/helm/blob/b2d5e41fc768882264af4c0162462fe9c6df21b4/pkg/action/upgrade.go#L336

And this copying leads to the issue where the tag `1.17` is taken up as part of values (overriding values), so chart's default values is not considered.

@jkrayl @Tarkleigh Could you check the values using `$ helm get values` for your old and new releases ? So that we can confirm that your issue is due to the values. Or else we need to raise the values issue that I just noticed as a different issue and look further into what's causing the problems in your chart upgrades

**Comment by karuppiah7890** [comment]
I think I found the issue with respect to install putting the computed values instead of just the user supplied values. During installation, while rendering values, we mutate the values in one place and then the same values gets stored in the release values. Here's the trace

https://github.com/helm/helm/blob/0f332b67f156eaff32cb67cec843260f7edba760/pkg/action/install.go#L195

https://github.com/helm/helm/blob/f94bac0643ad5d39c740c57c6c8ea6a4569a1db0/pkg/chartutil/values.go#L156

https://github.com/helm/helm/blob/b49db9e6e6fb1913a125b8cf9b31f81199c6753c/pkg/chartutil/coalesce.go#L40

https://github.com/helm/helm/blob/b49db9e6e6fb1913a125b8cf9b31f81199c6753c/pkg/chartutil/coalesce.go#L50

An example mutation:
https://github.com/helm/helm/blob/b49db9e6e6fb1913a125b8cf9b31f81199c6753c/pkg/chartutil/coalesce.go#L164

I tried fixing by copying values in `CoalesceValues` function like below and it worked.

```golang
func CoalesceValues(chrt *chart.Chart, vals map[string]interface{}) (Values, error) {
	currentVals := make(map[string]interface{})
	for key, value := range vals {
		currentVals[key] = value
	}
	if _, err := coalesce(chrt, currentVals); err != nil {
		return currentVals, err
	}
	return coalesceDeps(chrt, currentVals)
}
```

Gotta write tests for the function and fix it. I think any feature using this function has possible bugs if the value passed is used after the function call. Might have to write tests for all those too I guess? Not sure

**Comment by Tarkleigh** [comment]
@karuppiah7890 
I don't know if you still need it but here is the output of the get values. I tested with mongodb, using the chart versions 5.2.0 first and then upgrading to 7.2.7. I just used the default values and did not set any parameters or made any changes to the values.yaml

```
helm3 get values mongo --revision 1                                                                                       d047213@C02YJ316JGH6
affinity: {}
clusterDomain: cluster.local
image:
  debug: false
  pullPolicy: Always
  registry: docker.io
  repository: bitnami/mongodb
  tag: 4.0.3
livenessProbe:
  enabled: true
  failureThreshold: 6
  initialDelaySeconds: 30
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
metrics:
  enabled: false
  image:
    pullPolicy: IfNotPresent
    registry: docker.io
    repository: forekshub/percona-mongodb-exporter
    tag: latest
  podAnnotations:
    prometheus.io/port: "9216"
    prometheus.io/scrape: "true"
  serviceMonitor:
    additionalLabels: {}
    alerting:
      additionalLabels: {}
      rules: {}
    enabled: false
mongodbDisableSystemLog: false
mongodbEnableIPv6: true
mongodbExtraFlags: []
mongodbSystemLogVerbosity: 0
nodeSelector: {}
persistence:
  accessModes:
  - ReadWriteOnce
  annotations: {}
  enabled: true
  size: 8Gi
podAnnotations: {}
podLabels: {}
readinessProbe:
  enabled: true
  failureThreshold: 6
  initialDelaySeconds: 5
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
replicaSet:
  enabled: false
  name: rs0
  pdb:
    minAvailable:
      arbiter: 1
      primary: 1
      secondary: 1
  replicas:
    arbiter: 1
    secondary: 1
  useHostnames: true
resources: {}
securityContext:
  enabled: true
  fsGroup: 1001
  runAsUser: 1001
service:
  annotations: {}
  port: 27017
  type: ClusterIP
tolerations: []
usePassword: true
```
After the upgrade 
```
helm3 get values mongo                                                                                                    d047213@C02YJ316JGH6
affinity: {}
affinityArbiter: {}
clusterDomain: cluster.local
configmap: null
image:
  debug: false
  pullPolicy: Always
  registry: docker.io
  repository: bitnami/mongodb
  tag: 4.0.3
ingress:
  annotations: null
  certManager: false
  enabled: false
  hosts:
  - name: mongodb.local
    path: /
  secrets: null
  tls:
  - hosts:
    - mongodb.local
    secretName: mongodb.local-tls
initConfigMap: {}
livenessProbe:
  enabled: true
  failureThreshold: 6
  initialDelaySeconds: 30
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
metrics:
  enabled: false
  extraArgs: ""
  image:
    pullPolicy: IfNotPresent
    registry: docker.io
    repository: forekshub/percona-mongodb-exporter
    tag: latest
  livenessProbe:
    enabled: false
    failureThreshold: 3
    initialDelaySeconds: 15
    periodSeconds: 5
    successThreshold: 1
    timeoutSeconds: 5
  podAnnotations:
    prometheus.io/port: "9216"
    prometheus.io/scrape: "true"
  readinessProbe:
    enabled: false
    failureThreshold: 3
    initialDelaySeconds: 5
    periodSeconds: 5
    successThreshold: 1
    timeoutSeconds: 1
  serviceMonitor:
    additionalLabels: {}
    alerting:
      additionalLabels: {}
      rules: {}
    enabled: false
mongodbDirectoryPerDB: false
mongodbDisableSystemLog: false
mongodbEnableIPv6: true
mongodbExtraFlags: []
mongodbSystemLogVerbosity: 0
nodeSelector: {}
persistence:
  accessModes:
  - ReadWriteOnce
  annotations: {}
  enabled: true
  mountPath: /bitnami/mongodb
  size: 8Gi
  subPath: ""
podAnnotations: {}
podLabels: {}
readinessProbe:
  enabled: true
  failureThreshold: 6
  initialDelaySeconds: 5
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
replicaSet:
  enabled: false
  name: rs0
  pdb:
    enabled: true
    minAvailable:
      arbiter: 1
      primary: 1
      secondary: 1
  replicas:
    arbiter: 1
    secondary: 1
  useHostnames: true
resources: {}
resourcesArbiter: {}
securityContext:
  enabled: true
  fsGroup: 1001
  runAsUser: 1001
service:
  annotations: {}
  port: 27017
  type: ClusterIP
tolerations: []
updateStrategy:
  type: RollingUpdate
usePassword: true
volumePermissions:
  enabled: false
  image:
    pullPolicy: Always
    registry: docker.io
    repository: bitnami/minideb
    tag: stretch
  resources: {}
```
I hope that helps

**Comment by karuppiah7890** [comment]
Thanks, just by a glance, I think the values haven't changed and are the same ones as the first installation chart's default values yaml. I'm away from keyboard, but I'll check this and also fix it in code to make sure install doesn't do this. 

But to fix the installations in your release and @jkrayl 's release, I need to think of a way so that we don't have to completely uninstall and reinstall. But if that's possible, you can go ahead and do it once I fix the issue, raise PR and it gets merged. But without uninstalling, even with the fix (fix will be in install command flow), or without fix, we have to see what to do to fix it. Will think on it and put my thoughts here :)

**Comment by karuppiah7890** [comment]
@Tarkleigh actually the two values yamls are a bit different. I'm not sure. Gotta check if upgrade is also changing values yaml. But surely there are problems regarding the values being stored, that caused some of your issues as you said you never set any parameters. By the way, you also mentioned you didn't change the values.yaml, did you pass any values yaml using `--values` or `-f` which you didn't change during upgrade? Or did you mean you didn't change the chart's default values yaml?

**Comment by Tarkleigh** [comment]
@karuppiah7890 
I meant that I didn't change the chart's default values, I used it as-is for both installations. And at least for my case, a solution that involves uninstalling the MongoDB would be fine, as it is only a test instance.

**Comment by karuppiah7890** [comment]
Cool, thanks :) then for now I'll spend sometime writing tests and fixing
the bug 😄 what about you @jkaryl ?

**Comment by jkrayl** [comment]
@karuppiah7890 uninstalling and reinstalling is fine with me :)

**Comment by thomastaylor312** [comment]
Fixed in #6430

**Comment by andrewnazarov** [comment]
I've experienced this twice with v3.0.0. Image tags haven't been updated even though the releases was installed successfully.
